library/bignum.c

```diff
@@ -2280,6 +2280,21 @@ int mbedtls_mpi_gen_prime( mbedtls_mpi *X, size_t nbits, int flags,
                 MBEDTLS_MPI_CHK( mbedtls_mpi_add_int( &Y, &Y, 6  ) );
             }
         }
+        /* check for TRNG vailidity if flag is set (Note: this has a false
+        positive probability of %1). if more than 5 * nlen/2
+        (nlen/2 is the size of the random buffer in bits) generations fail
+        return an error indicating a possible issue with TRNG.
+        */
+        if( ( flags & MBEDTLS_MPI_GEN_PRIME_CHECK_TRNG ) ==
```

Comment:
This check is at the end of the loop. If the bad RNG happens to put the code goes through the `continue` path above, this check will not be executed and so you still have an infinite loop.

Reply:
I haven't checked FIPS 186-4, but in general if FIPS and common sense disagree we should go with common sense. Does FIPS say that you _must not_ avoid an infinite loop there, or does it just not say anything about this case?

Reply:
the pseudo code from fips is here:
https://nvlpubs.nist.gov/nistpubs/fips/nist.fips.186-4.pdf#page=64

fips definitely doesn't say you are not allowed to avoid an infinite loop, but the loop avoidance they offer doesn't cover this case (see in in the section linked above B3.3 steps 4.1-4.7). I'm not sure if this is the best possible approach , but there may be a good reason why they don't seem to handle this.

I think maybe they don't enforce the check for random numbers which are too small because the probability bounds on it are completely different than on large numbers that fail the primeness checks. maybe @yanesca  can help here.
I think enforcing this new kind of limit requires a new probability bound not provided in this fips section.
Note that our implementation is a bit different than the one in the fips pseudo code(specifically the primality checks) so that the checks are in the spirit of the fips check, but not exactly the same.

Reply:
I don't see why we couldn't introduce a check where @gilles-peskine-arm proposes it. It would be the cleanest to use a different counter, also we could adjust the FIPS bounds accordingly. In the either case we would need to thoroughly document how we arrived at the bound. If the input is uniform then the chance of entering the loop there is around 71%. So if we allow 30 iterations then the probability of false positives will be around 0.00003%. If this is still too high we can allow more.

On the other hand the goal of the PR is to "increase adherence to FIPS" and not to "devise and implement sensible tests for RNG failure" so I feel it being slightly out of scope.

But again it is a small change and makes sense so I don't see why we shouldn't do it. (I think that in FIPS they probably omitted it because an RNG failure causing an infinite loop there is so trivial that a health tests in a NIST approved TRNG would catch it.)

Reply:
so just to clarify, basically there is an agreement to add an additional check (in addition to the fips one) that will check the number of retries for the "too small" case (which normally just retries). reaching this case over 30 times per call will trigger an error (same error as the fips bound?).  can I proceed with this proposal?

I will probably need a more formal documentation of the new check (can't fall back on fips for this one). @yanesca any chance you could help me with the formal explanation?

Reply:
Thanks @yanesca , I'll start working on this. @gilles-peskine-arm please ping if you have any objections.

Reply:
I have now checked FIPS 186-4 and the original version of the PR was as close as it gets to the FIPS iteration count as it gets without doing some significant refactoring between prime generation and RSA key generation. Furthermore, now that I have a better view of what this tests, I think this is a compliance issue and not a functional issue, and I no longer think that we should protect the other failure causes. If we do, then I think we need a separate iteration counter, because they have wildly differing probabilities of failure.

I don't like adding yet another function just to be a bit closer to FIPS. Either we go for FIPS compliance (which means more reengineering of the code than we're doing now) or we don't (in which case I question the purpose of this pull request).